src/bootstrap/tool.rs

```diff
@@ -76,159 +73,22 @@ impl Step for ToolBuild {
-                for (id, cur, prev) in different {
-                    eprintln!("  {}", id);
-                    let cur_features: HashSet<_> = cur.2.into_iter().collect();
-                    let prev_features: HashSet<_> = prev.2.into_iter().collect();
-                    eprintln!(
-                        "    `{}` additionally enabled features {:?} at {:?}",
-                        cur.0,
-                        &cur_features - &prev_features,
-                        cur.1
-                    );
-                    eprintln!(
-                        "    `{}` additionally enabled features {:?} at {:?}",
-                        prev.0,
-                        &prev_features - &cur_features,
-                        prev.1
-                    );
-                }
-            }
-            eprintln!();
-            eprintln!(
-                "to fix this you will probably want to edit the local \
-                      src/tools/rustc-workspace-hack/Cargo.toml crate, as \
-                      that will update the dependency graph to ensure that \
-                      these crates all share the same feature set"
-            );
-            panic!("tools should not compile multiple copies of the same crate");
+        // HACK(#82501): on Windows, the tools directory gets added to PATH when running tests, and
+        // compiletest confuses HTML tidy with the in-tree tidy. Name the in-tree tidy something
+        // different so the problem doesn't come up.
+        if tool == "tidy" {
+            tool = "rust-tidy";
         }
 
-        builder.save_toolstate(
-            tool,
-            if is_expected { ToolState::TestFail } else { ToolState::BuildFail },
-        );
-
-        if !is_expected {
-            if !is_optional_tool {
-                crate::detail_exit(1);
-            } else {
-                None
-            }
-        } else {
-            // HACK(#82501): on Windows, the tools directory gets added to PATH when running tests, and
-            // compiletest confuses HTML tidy with the in-tree tidy. Name the in-tree tidy something
-            // different so the problem doesn't come up.
-            if tool == "tidy" {
-                tool = "rust-tidy";
-            }
-            let cargo_out = builder.cargo_out(compiler, self.mode, target).join(exe(tool, target));
-            let bin = builder.tools_dir(compiler).join(exe(tool, target));
-            builder.copy(&cargo_out, &bin);
-            Some(bin)
-        }
+        // Streamlined building all tools mandatorily
+        let cargo_out = builder.cargo_out(compiler, self.mode, target).join(exe(tool, target));
+        let cargo_stamp = builder.cargo_out(compiler, self.mode, target).join(".libtool.stamp");
+        let bin = builder.tools_dir(compiler).join(exe(tool, target));
+        compile::run_cargo(builder, cargo, vec![], &cargo_stamp, vec![], false, false);
```

Comment:
A small optimization:
```suggestion
        let out_dir = builder.cargo_out(compiler, self.mode, target);
        let cargo_out = out_dir.join(exe(tool, target));
        let bootstrap_tool_stamp = out_dir.join(".bootstrap-tool.stamp");
        let bin = builder.tools_dir(compiler).join(exe(tool, target));

        // Streamlined building all tools mandatorily
        compile::run_cargo(builder, cargo, vec![], &bootstrap_tool_stamp, vec![], false, false);
```